### PR TITLE
BETA version of the Storybook Multi Input 

### PR DESCRIPTION
**Pull request by Ridhwana**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

I created an MVP of the Multi Input Component. It does not yet cover all of the functionality required, but I'd like to merge this one in and then build upon it.

I would still need to add:
- validation 
- allow the regex for an input to be passed through as a prob
- add tests 
- add documentation
- delete an item using keyboard navigation (accessibility)

## Related Tickets & Documents

- Related Issue #https://github.com/forem/forem/issues/17577

## QA Instructions, Screenshots, Recordings

- Locally run yarn install && yarn storybook
- Navigate to the BETA > MultiInput story
- Play around with adding and removing items (email addresses).

Demo Video:
https://www.loom.com/share/4f7a9e49bc3e4739a4a7e20453389488

It should look somewhat like this, however in the next PR I will focus on polishing it up.
Screenshot 2022-06-06 at 15 14 46

### UI accessibility concerns?

I tested keyboard navigation and voice over, but could have potentially missed a few things out.
 
## Added/updated tests?

- [ ] Yes
- [x] No, and this is why: I'll add tests once I polish up the BETA version
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [x] This change does not need to be communicated right now

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/assets/stylesheets/components/multi-input.scss
Comment (Ridhwana) [inline app/assets/stylesheets/components/multi-input.scss:5]: `c-input--multi__input` this classname may look a bit weird but was struggling to come up with an alternative. Open to thoughts and suggestions.

File: app/javascript/crayons/MultiInput/MultiInput.jsx
Comment (Ridhwana) [inline app/javascript/crayons/MultiInput/MultiInput.jsx:9]: I've added some specific todo notes in this file that I will tackle in a subsequent PR.

File: app/javascript/crayons/MultiInput/MultiInput.jsx
Comment (Ridhwana) [inline app/javascript/crayons/MultiInput/MultiInput.jsx:74]: When testing voice over on this component, I realized that I did not add an event handle for KEYS.DELETE as it seemed a little more complicated so I'll tackle it on the next PR. I _think_ everything else should be accessible based on my testing but would appreciate another set of eyes please.

File: app/javascript/crayons/MultiInput/__stories__/MultiInput.stories.jsx
Comment (Ridhwana) [inline app/javascript/crayons/MultiInput/__stories__/MultiInput.stories.jsx:26]: In the designs we do not have a placeholder "Add an email address..." but I thought it may be needed for accessibility reasons and also to make the UI easier to understand. Looking for thoughts here.

File: app/javascript/crayons/MultiInput/__stories__/MultiInput.mdx
Comment (Ridhwana) [inline app/javascript/crayons/MultiInput/__stories__/MultiInput.mdx:1]: I'll be filling out the documentation once I've polished the component.

File: app/javascript/crayons/MultiInput/__stories__/MultiInput.stories.jsx
Comment (juliannatetreault) [reply app/javascript/crayons/MultiInput/__stories__/MultiInput.stories.jsx:26]: I like the addition of this for accessibility purposes (it looks like we also do this for the tag autocomplete). I also think that "Add an email address..." is consistent with what we have elsewhere!

File: app/javascript/crayons/MultiInput/__stories__/MultiInput.mdx
Comment (juliannatetreault) [inline app/javascript/crayons/MultiInput/__stories__/MultiInput.mdx:1]: Quick, non-blocking suggestion: let's capitalize the "i" here and hyphenate the word! I'm okay with holding off on changing this until when the rest of the documentation is tackled, so that this tiny comment doesn't hold up the PR. 😊 
```suggestion
# Multi-Input
```

**Review by juliannatetreault — APPROVED** [review]
This is ✨ amazing ✨ !!

File: app/javascript/crayons/MultiInput/MultiInput.jsx
Comment (aitchiss) [inline app/javascript/crayons/MultiInput/MultiInput.jsx:32]: I'm not sure we want the entry to be selected on `Enter` press 🤔  We allow this in the tag multi-select because it follows the combobox pattern where you scroll down with the keyboard arrows to select an option, but for this `MultiInput` we are following default input behaviour (which would be to submit the form on Enter press if it's the last field in the form).

IDK like I say I'm not sure - what do you think?

File: app/javascript/crayons/MultiInput/MultiInput.jsx
Comment (aitchiss) [inline app/javascript/crayons/MultiInput/MultiInput.jsx:71]:
```suggestion
```

I don't think we need this, since "false" is the default value

File: app/javascript/crayons/MultiInput/MultiInput.jsx
Comment (aitchiss) [inline app/javascript/crayons/MultiInput/MultiInput.jsx:67]: Are we also going to add the "edit in place" feature that we have on the tag autocomplete selections - so when you click on the text rather than the X button, the input moves to that position and is pre-filled with the item text to begin editing it?

If we don't need it for this feature, then we could remove some code (like this group and the `style="order: 1"`)

File: app/javascript/crayons/MultiInput/MultiInput.jsx
Comment (aitchiss) [inline app/javascript/crayons/MultiInput/MultiInput.jsx:68]: I think we could use the `Pill` Preact component which wraps this functionality 😄

**Review by aitchiss — COMMENTED** [review]
Woohoo this is exciting! It's looking great 🤩  🔥 

In terms of a11y testing, I think the main thing missing is a confirmation announcement when an item is added or removed from the list. In the `MultiSelectAutocomplete` we achieve that with a hidden `aria-live` region, but depending whether or not we are adding the "Edit in place" functionality to this feature, there could potentially be a simpler path forward. 

There's also an issue at the moment with keyboard access to the pill items, as I can tab to them but not press `Enter` for the delete action (as the click handler is on the SVG rather than the button). Again, depending on whether we add the "edit" functionality, we'll need to make some changes to the selected items markup.

These things can definitely be tackled in subsequent PRs - just making sure we capture them as tasks!

File: app/javascript/crayons/MultiInput/MultiInput.jsx
Comment (Ridhwana) [reply app/javascript/crayons/MultiInput/MultiInput.jsx:32]: I used Gmail as an inspiration here and looked at the way that adding email addresses when composing a mail on gmail. They seem to select the entry on enter, space, comma and tab. Aside note: I didn't implement the select on tab but would be curious to get your thoughts? 

To me it feels intuitive and quite nice to press enter and have it selected (i.e. move to a pill). However, from an accessibility point of view, if we think it's conflicting with natural behaviour the we should remove it. 

TLDR;I think its nice for the entry to be selected on Enter but if we think its conflicting with natural accessible behaviour then I can remove it.

Comment (aitchiss) [reply app/javascript/crayons/MultiInput/MultiInput.jsx:32]: That's good context! I think on balance let's leave it then because I can also see maintaining the default behaviour having potential for confusion (we don't want someone to think it works in the Gmail way, and then accidentally submit the form!)

I don't think we need to do anything on Tab, because Tab key moves the user out of the input and triggers the `onBlur` which selects the current text anyway 🙌

File: app/javascript/crayons/MultiInput/MultiInput.jsx
Comment (Ridhwana) [reply app/javascript/crayons/MultiInput/MultiInput.jsx:67]: Just confirmed with Anuj now, and based on [the conversation](https://forem-team.slack.com/archives/C01S4QZUDNC/p1654594483016859) I'm going to leave it in for now, until I tackle the Edit functionality in subsequent PRs.

**Comment by Ridhwana** [comment]
> Woohoo this is exciting! It's looking great 🤩 🔥
> 
> In terms of a11y testing, I think the main thing missing is a confirmation announcement when an item is added or removed from the list. In the `MultiSelectAutocomplete` we achieve that with a hidden `aria-live` region, but depending whether or not we are adding the "Edit in place" functionality to this feature, there could potentially be a simpler path forward.
> 
> There's also an issue at the moment with keyboard access to the pill items, as I can tab to them but not press `Enter` for the delete action (as the click handler is on the SVG rather than the button). Again, depending on whether we add the "edit" functionality, we'll need to make some changes to the selected items markup.
> 
> These things can definitely be tackled in subsequent PRs - just making sure we capture them as tasks!

Thanks so much for outlining these @aitchiss and bringing up the accessibility questions and thoughts. I really appreciate it!✨ I've noted the above and those that that I'll tackle in a subsequent PR. If you think there's anything else that I should address in this PR before we can merge, let me know!

**Comment by aitchiss** [comment]
> Thanks so much for outlining these @aitchiss and bringing up the accessibility questions and thoughts. I really appreciate it!✨ I've noted the above and those that that I'll tackle in a subsequent PR. If you think there's anything else that I should address in this PR before we can merge, let me know!

All sounds great to me - let's get this merged 🚀

**Pull request merged**
